### PR TITLE
Add project generation request statistics support

### LINKED ISSUE #53: Enbale tracking and collecting the data usage of Azure Spring Initializr project generation
### Context
Collect the number of generated projects, the metadata of generated projects (JDK version, Spring versions, Spring Cloud Azure versions), and the success rates of generating projects. 

What's the problem to solve?
- Filter the project generation request log to Log Analysis
- Parse the JSON format log to Kusto for the BI report
- Reduce dependencies on additional resources or tools

Candidate solutions for ETL:
The log in Azure Spring Cloud can be sent to the following destinations
- ❌**Log Analysis (ASC service level)**
  The JSON format log can not be directly passed to the structured log into a Kusto table, then the data should be exported into Storage Account first, then Kusto does ingestion from the Storage Account.

- **Storage Account**
 2 subs solutions when using Storage Account:
   - ❌**Send log to Storage Account in ASC service level**
     The same question to solve is to parse the JSON log format before ingestion of the data into Kusto.
  
   - ❌**Send log into Storage directly in the application of the ASC**
     This solution can solve the JSON data parsing, but another question is to control the right size to upload the blob to trigger the blob-created event which the Kusto will listen to the event in a special container, and lots of log data will be merged first, then move to the special container. The blob files of the subfolder will not be listened to transform data, this will bring the additional process to do the data moving.
     Also, the blob files should regularly be maintained.

- **Event Hubs**
 2 subs solutions when using Event Hub:
   - ❌**Send log to Event Hub in ASC service level**
     The same question to solve is to parse the JSON log format before Kusto consumes the event for ingestion.
  
   - ✅**Send log into Event Hub directly in the application of the ASC**
     This solution can solve the JSON data parsing, and the data in Kusto does not need to change anymore.

### Goal
- Collect the data of the project generation which contains the success and failure request
- Daily and monthly generated project number report
  - Can filter by JDK version, Spring Boot versions, language

### Solutions
- Output project generation logs to 2 EventHubs when enabling the EventHub stats
- Kusto will ingest the log via consuming the structured data from a special Event Hub
- Generate BI report from the Kutso data.

### PR DESCRIPTION
**Pull request by moarychan**
Fixes https://github.com/Azure/azure-spring-initializr/issues/53

Add log for project request usage, we use the Event hubs to transform the request event and finally save the data into one table in Azure Data Explorer, this feature is disabled by default.
Use one log structure to record and send the event to the Event Hubs, this event will contain the success and failure request data.

### CODE REVIEW

File: start-site/src/main/resources/application.yml [hunk 1]
```diff
@@ -37,6 +37,12 @@ spring:
       cache:
         use-last-modified: false
 
+  cloud:
+    azure:
+      eventhubs:
+        event-hub-name: ${INITIALIZR_SUCCESS_EVENT_HUB}
+        connection-string: ${EVENT_HUBS_INITIALIZR_CONNECTION_STRING}
```
Comment (stliu) [inline start-site/src/main/resources/application.yml:44]: we should use MSI and make this application **credential-free**

File: start-site/pom.xml [hunk 2]
```diff
@@ -47,7 +47,10 @@
 			<artifactId>start-client</artifactId>
 			<scope>runtime</scope>
 		</dependency>
-
+    <dependency>
```
Comment (backwind1233) [inline start-site/pom.xml:50]: wrong indent

File: start-site/src/main/resources/application.yml [hunk 3]
```diff
@@ -37,6 +38,14 @@ spring:
       cache:
         use-last-modified: false
 
+  cloud:
```
Comment (backwind1233) [inline start-site/src/main/resources/application.yml:41]: can we move this into `initializr-extention` module?

File: initializr-extension/src/main/java/com/azure/spring/initializr/autoconfigure/ExtendInitializrAutoConfiguration.java [hunk 4]
```diff
@@ -85,6 +86,8 @@
 @AutoConfigureAfter({ JacksonAutoConfiguration.class, RestTemplateAutoConfiguration.class })
 public class ExtendInitializrAutoConfiguration {
 
+    private static final Logger LOGGER = org.slf4j.LoggerFactory.getLogger(ExtendInitializrAutoConfiguration.class);
```
Comment (saragluna) [inline initializr-extension/src/main/java/com/azure/spring/initializr/autoconfigure/ExtendInitializrAutoConfiguration.java:89]: Why do we have the package name here for LoggerFactory?

File: initializr-extension/src/main/java/com/azure/spring/initializr/autoconfigure/ExtendInitializrProperties.java [hunk 5]
```diff
@@ -16,6 +16,9 @@ public class ExtendInitializrProperties {
 
     private static final Logger LOGGER = org.slf4j.LoggerFactory.getLogger(ExtendInitializrProperties.class);
 
+    @JsonIgnore
```
Comment (saragluna) [inline initializr-extension/src/main/java/com/azure/spring/initializr/autoconfigure/ExtendInitializrProperties.java:19]: Same here for the `LoggerFactory`

File: initializr-extension/src/main/java/com/azure/spring/initializr/autoconfigure/ExtendInitializrAutoConfiguration.java [hunk 6]
```diff
@@ -93,8 +96,21 @@ public ProjectDirectoryFactory projectDirectoryFactory() {
...
-    public ProjectGenerationListener projectGenerationListener() {
-        return new ProjectGenerationListener();
+    public ProjectGenerationListener projectGenerationListener(ObjectMapper objectMapper,
+                                                               ObjectProvider<StatProjectGenerationLogProcessor> processors) {
+        return new ProjectGenerationListener(objectMapper, processors.getIfUnique());
+    }
+
+    @Bean
+    @ConditionalOnMissingBean(StatProjectGenerationLogProcessor.class)
+    @ConditionalOnProperty(value = "extend.initializr.stats.eventhub.enabled", havingValue = "true")
```
Comment (saragluna) [inline initializr-extension/src/main/java/com/azure/spring/initializr/autoconfigure/ExtendInitializrAutoConfiguration.java:106]: eventhubs

File: initializr-extension/src/main/java/com/azure/spring/initializr/autoconfigure/ExtendInitializrAutoConfiguration.java [hunk 7]
```diff
@@ -93,8 +96,21 @@ public ProjectDirectoryFactory projectDirectoryFactory() {
...
+    public StatProjectGenerationLogProcessor eventHubProcessor(ExtendInitializrProperties properties,
```
Comment (saragluna) [inline initializr-extension/src/main/java/com/azure/spring/initializr/autoconfigure/ExtendInitializrAutoConfiguration.java:107]: Please change the bean name to match its usage

File: initializr-extension/src/main/java/com/azure/spring/initializr/autoconfigure/ExtendInitializrProperties.java [hunk 8]
```diff
@@ -16,6 +16,9 @@ public class ExtendInitializrProperties {
...
+    @NestedConfigurationProperty
+    private StatsProperties stats = new StatsProperties();
 
     @JsonIgnore
 //    @NestedConfigurationProperty
```
Comment (saragluna) [inline initializr-extension/src/main/java/com/azure/spring/initializr/autoconfigure/ExtendInitializrProperties.java:24]: Please help remove this commented out line

File: initializr-extension/src/main/java/com/azure/spring/initializr/autoconfigure/StatsProperties.java [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+package com.azure.spring.initializr.autoconfigure;
+
+import org.springframework.boot.context.properties.NestedConfigurationProperty;
+
+/**
+ * Statistics-related properties.
+ */
+public class StatsProperties {
+
+	@NestedConfigurationProperty
```
Comment (saragluna) [inline initializr-extension/src/main/java/com/azure/spring/initializr/autoconfigure/StatsProperties.java:26]: No need for `@NestedConfigurationProperty`; when a class is a static internal class, the `@NestedConfigurationProperty` can be omitted

File: initializr-extension/src/main/java/com/azure/spring/initializr/autoconfigure/StatsProperties.java [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+	private final Eventhub eventhub = new Eventhub();
+
+    public Eventhub getEventhub() {
+        return eventhub;
+    }
+
+    /**
+	 * EventHub configuration.
+	 */
+	public static final class Eventhub {
```
Comment (saragluna) [inline initializr-extension/src/main/java/com/azure/spring/initializr/autoconfigure/StatsProperties.java:36]: The indentation doesn't seem correct.

Comment (saragluna) [inline initializr-extension/src/main/java/com/azure/spring/initializr/autoconfigure/StatsProperties.java:36]: Should we call it `Eventhubs`?

File: initializr-extension/src/main/java/com/azure/spring/initializr/web/project/EventHubStatProjectGenerationLogProcessor.java [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+import com.azure.messaging.eventhubs.EventHubProducerClient;
+import com.azure.spring.initializr.autoconfigure.StatsProperties;
+import org.slf4j.Logger;
+
+import java.util.Arrays;
+
+/**
+ * Using Event Hub to distribute statistical project generation logs.
+ */
+public class EventHubStatProjectGenerationLogProcessor implements StatProjectGenerationLogProcessor {
```
Comment (saragluna) [inline initializr-extension/src/main/java/com/azure/spring/initializr/web/project/EventHubStatProjectGenerationLogProcessor.java:15]: EventHubs

File: initializr-extension/src/main/java/com/azure/spring/initializr/web/project/GenerationEventLogConverter.java [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+import io.spring.initializr.web.project.ProjectRequestEvent;
+import org.springframework.beans.BeanUtils;
+import org.springframework.core.convert.converter.Converter;
+
+import java.time.Instant;
+import java.time.LocalDateTime;
+import java.time.ZoneId;
+import java.util.Optional;
+
+public class GenerationEventLogConverter implements Converter<ProjectRequestEvent, GenerationEventLog> {
```
Comment (saragluna) [inline initializr-extension/src/main/java/com/azure/spring/initializr/web/project/GenerationEventLogConverter.java:12]: should this be a final class if with a private ctor?

File: initializr-extension/src/main/java/com/azure/spring/initializr/web/project/ProjectFailedEventLogConverter.java [hunk 13]
```diff
@@ -0,0 +1,25 @@
+package com.azure.spring.initializr.web.project;
+
+import io.spring.initializr.web.project.ProjectFailedEvent;
+import org.springframework.beans.BeanUtils;
+import org.springframework.core.convert.converter.Converter;
+
+public class ProjectFailedEventLogConverter implements Converter<ProjectFailedEvent, GenerationFailedEventLog> {
```
Comment (saragluna) [inline initializr-extension/src/main/java/com/azure/spring/initializr/web/project/ProjectFailedEventLogConverter.java:7]: Should this class be called `GenerationFailedEventLogConverter`

File: initializr-extension/src/main/java/com/azure/spring/initializr/web/project/StatProjectGenerationLogProcessor.java [hunk 14]
```diff
@@ -0,0 +1,11 @@
+package com.azure.spring.initializr.web.project;
+
+/**
+ * Statistical project generation log processor interface.
+ */
+public interface StatProjectGenerationLogProcessor {
```
Comment (saragluna) [inline initializr-extension/src/main/java/com/azure/spring/initializr/web/project/StatProjectGenerationLogProcessor.java:6]: How about changing the name to `ProjectGenerationStatisticsProcessor`?

**Review by saragluna — CHANGES_REQUESTED** [review]

File: initializr-extension/src/main/java/com/azure/spring/initializr/web/project/EventHubsProjectGenerationStatisticsProcessor.java [hunk 15]
```diff
@@ -0,0 +1,30 @@
...
+    private static final Logger LOGGER = LoggerFactory.getLogger(EventHubsProjectGenerationStatisticsProcessor.class);
+    private final EventHubProducerClient producerClient;
+
+    public EventHubsProjectGenerationStatisticsProcessor(EventHubProducerClient producerClient) {
+        LOGGER.info("Project generation listener uses the Event Hubs processor implementation to distribute logs.");
+        this.producerClient = producerClient;
+    }
+
+    @Override
+    public void sendRequest(String json) {
```
Comment (stliu) [inline initializr-extension/src/main/java/com/azure/spring/initializr/web/project/EventHubsProjectGenerationStatisticsProcessor.java:25]: don't like this method naming, it couples with impl

File: initializr-extension/src/main/java/com/azure/spring/initializr/web/project/ProjectGenerationStatisticsProcessor.java [hunk 16]
```diff
@@ -0,0 +1,9 @@
+package com.azure.spring.initializr.web.project;
+
+/**
+ * Statistical project generation log processor interface.
+ */
+public interface ProjectGenerationStatisticsProcessor {
+
+    void sendRequest(String json);
```
[the remark this answers is not in the document]
Comment (moarychan) [reply initializr-extension/src/main/java/com/azure/spring/initializr/web/project/ProjectGenerationStatisticsProcessor.java:8]: Indeed, I will refactor the design here.

**Comment by moarychan** [comment]
Closing it due to dropping from the plan.

**Pull request closed without merging**